src/proof/fischlin.rs

```diff
@@ -0,0 +1,356 @@
+
+        let challenges = challenges_and_responses.map(|(challenge, _)| challenge);
+
+        let responses = challenges_and_responses.map(|(_, response)| response);
+
+        let maurer_proof = super::Proof::new(statement_masks_values, responses);
+
+        let uc_proof = Self {
+            maurer_proof,
+            challenges,
+        };
+
+        Ok((uc_proof, statement))
+    }
+
+    /// Verify a Universally Composable (UC) Extractable Maurer zero-knowledge claim via Fischlin's transform.
+    /// Implements [Chen and Lindell (2024)](https://eprint.iacr.org/2024/526.pdf), sections 2.2, 2.3.
+    pub fn verify(
+        &self,
+        protocol_context: &ProtocolContext,
+        language_public_parameters: &Language::PublicParameters,
+        statement: Language::StatementSpaceGroupElement,
+    ) -> Result<()> {
+        // $common-h$
+        let common_hash = Self::compute_common_hash(
+            protocol_context,
+            language_public_parameters,
+            &statement,
+            &self.maurer_proof.statement_masks,
+        )?;
+
+        let hash_checks: Vec<_> = self
+            .challenges
+            .into_iter()
+            .zip(self.maurer_proof.responses)
+            .enumerate()
+            .map(|(i, (challenge, response))| {
+                Self::hash_hits_target(common_hash, i, challenge, &response)
+            })
+            .collect();
+
+        if hash_checks.iter().any(|res| res.is_err()) {
+            return Err(Error::InternalError);
+        } else if hash_checks
+            .into_iter()
+            .map(|res| res.unwrap())
+            .any(|is_target| !is_target)
+        {
+            return Err(proof::Error::ProofVerification)?;
+        }
+
+        self.maurer_proof.verify_inner(
+            self.challenges.map(|challenge| vec![challenge]),
+            language_public_parameters,
+            vec![statement],
+        )
+    }
+
+    /// Compute $common-h$ as a full hash with output length $2\cdot\kappa_c$.
+    fn compute_common_hash(
+        protocol_context: &ProtocolContext,
+        language_public_parameters: &Language::PublicParameters,
+        statement: &Language::StatementSpaceGroupElement,
+        statement_masks_values: &[group::Value<Language::StatementSpaceGroupElement>; REPETITIONS],
+    ) -> Result<<ComputationalSecuritySizedNumber as Concat>::Output> {
+        let mut transcript =
+            super::Proof::<REPETITIONS, Language, ProtocolContext>::setup_transcript(
+                protocol_context,
+                language_public_parameters,
+                vec![statement.value()],
+                statement_masks_values,
+            )?;
+
+        Ok(transcript.challenge(b"common hash"))
+    }
+
+    /// Performs the Fischlin transformation check that $h_i = H(common-h, i, e_i, z_i)$ hits the target,
+    /// i.e. it starts with `target_bits` $b$ zeros.
+    fn hash_hits_target(
```

Comment:
```suggestion
    fn is_target(
```
@dolevmu which name is more suitable?

Reply:
I like `hash_hits_target`. Also your input to this function is the input to the hash H_b, so you don't ask if the input **is** the target, but if the **hash of it is** the target. So at least `hash_is_target` but hits sounds better.